utils: Tracing using the retry-chain-node

Lazin opened the pull request:
This PR adds few things:
- `execution_monitor` class which implements a periodic liveness check;
- `retry_chain_context` is a mechanism that can be used to collect trace logs;

We have the following problem. In case of any liveness issue (deadlock) it's difficult to debug the `ntp_archiver` because the system runs on info level so the trace logs from the moment when the issue occurred are gone. 

The proposed solution to this is to improve observability of the system. Instead of collecting trace logs all the time we can collect them in memory. When something interesting happens we can dump the diagnostic information and include all trace logs that we collected. This guarantees that the traces will be visible.

The `retry_chain_context` solves this issue. The `retry_chain_node` is modified to take the context as a parameter. When any information is logged using the `retry_chain_logger` the data is written into the context. These traces are stored in memory using limited amount of memory (configurable). The code can iterate through traces or discard them.

The last but not least is the `execution_monitor`. The monitor is a simple tool that takes a list of pointers to `retry_chain_context` objects and a pointer to `seastar::gate` and two callbacks. It checks the contexts periodically. If the contexts were not accessed long enough it detects a stall and invokes a corresponding callback. It also checks the gate. If the gate is closed before the execution monitor is stopped it detects an unexpected exits and invokes a corresponding callback.

The code in the `ntp_archiver` is modified to use all these mechanisms to detect stalls and log the diagnostics.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* none

Lazin committed:
utils: Update retry_chain_node

Add 'retry_chain_context' class. The context stores the reference to the
abort source so it could be used to abort the computation.

Also, the retry_chain_context can accumulate trace logs using limited
amount of memory. The traces could be retrieved. The context has a
'reset' method that removes all accumulated traces.

The 'retry_chain_logger' can interact with the context to save the
traces. The idea is that we can dump all trace logs in case of stall or
unexpected failure.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin commented:
@nvartolomei this is the idea that we discussed the other day, the `retry_chain_node` is used to collect all traces generated using the `retry_chain_logger`.

mmaslankaprv commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,307 @@
+    }
+
+private:
+    seastar::future<> lock(seastar::sstring checkpoint_name) {
```
can we use `std::string_view` ? also it would be great to use `vlog::file_line::current()` as the checkpoint name so caller doesn't have to bother with naming checkpoints

Lazin replied:
Are you suggesting to store `std::string_view` and make the caller responsible for the lifetime?

Lazin replied:
The mutex was moved into a separate PR now - https://github.com/redpanda-data/redpanda/pull/26381
It no longer uses strings. It uses `vlog::file_line` instead so almost no runtime overhead.

Lazin committed:
ssx: Add execution_monitor class

The monitor is checking a set of site periodically and can detects
stalls and unexpected shutdown of the process.

The detector is using a periodic timer to check a set of pre-existing
callsties. The callsite is a simple data structure that contains a time
of the last checkpoint. There is a finite set of callsites.

The process invokes the 'checkpoint' method of the callsite while it
progresses.

There are also shutdown callsites. These are callsites which are reached
when the process is being stopped.

The execution monitor checks normal callsites and shutdown callsites
every stall-check-interval.

If all normal callsites weren't checkpointed within the stall interval
the stall is detected. If any shutdown callsite was checkpointed before
the 'expect_shutdown' was called the unexpected shutdown is detected.

In both cases the monitor invokes the callback provided by the user.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

dotnwat reviewed (commented):
should this be 3 or 4 separate PRs?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/67107-->
<!--8c728738-a157-4f46-a070-3690c03c8521-->
<!--6ca7bf9b-d0f5-4a8d-a0d0-a142c0ccc946-->
<!--42bfa24b-d976-4c46-9c52-f0c057230382-->
<!--d23093e8-580d-46dd-a903-ebe9dc30b6f6-->
<!--d079dcfc-41d5-4212-9f78-1022a82e22e5-->
<!--a748349c-48e4-465d-88c1-a4e547025754-->
<!--3e154dce-687c-4d90-aabf-bb4d306c4d9c-->
<!--title-->
## Retry command for [Build#67107](https://buildkite.com/redpanda/redpanda/builds/67107)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/read_replica_e2e_test.py::TestReadReplicaService.test_simple_end_to_end@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"partition_count":10}
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete@{"cloud_storage_type":2,"retention_type":"retention.ms"}
tests/rptest/tests/read_replica_e2e_test.py::TestReadReplicaService.test_writes_forbidden@{"cloud_storage_type":1,"partition_count":10}
tests/rptest/tests/read_replica_e2e_test.py::TestReadReplicaService.test_simple_end_to_end@{"cloud_storage_type_and_url_style":[1,"path"],"partition_count":10}
tests/rptest/tests/read_replica_e2e_test.py::TestReadReplicaService.test_identical_hwms@{"cloud_storage_type":1,"partition_count":5}
tests/rptest/tests/timequery_test.py::TimeQueryTest.test_timequery@{"batch_cache":false,"cloud_storage":true,"spillover":true}
tests/rptest/tests/read_replica_e2e_test.py::TestReadReplicaService.test_simple_end_to_end@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"partition_count":10}
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete@{"cloud_storage_type":1,"retention_type":"retention.ms"}
tests/rptest/tests/read_replica_e2e_test.py::TestReadReplicaService.test_identical_lwms_after_delete_records@{"cloud_storage_type":1,"partition_count":5}
tests/rptest/tests/timequery_test.py::TimeQueryTest.test_timequery_with_spillover_gc_delayed
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage@{"cleanup_policy":"compact,delete"}
tests/rptest/tests/read_replica_e2e_test.py::TestReadReplicaService.test_partition_movement@{"partition_count":10}
tests/rptest/tests/read_replica_e2e_test.py::ReadReplicasUpgradeTest.test_upgrades@{"cloud_storage_type":1}
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete@{"cloud_storage_type":2,"retention_type":"retention.bytes"}
tests/rptest/tests/read_replica_e2e_test.py::TestReadReplicaService.test_writes_forbidden@{"cloud_storage_type":2,"partition_count":10}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTimelyGCTest.test_retention_with_node_failures@{"cloud_storage_type":1}
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves@{"cleanup_policy":"delete"}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--2dc47cef-2332-471c-be82-f252a34e2140-->
<!--TEST RESULTS-->
<!--75eded2a-81d6-4f86-9400-de6b07c2a308-->
<!--TEST RESULTS-->
<!--696da197-4b47-4eff-9e47-e18008464323-->
<!--TEST RESULTS-->
<!--99a9b625-9c46-404f-b695-f519eb7a4393-->
<!--TEST RESULTS-->
<!--6cf4163b-6675-4099-a6ed-74e46699a617-->
<!--TEST RESULTS-->
<!--b87cb494-817f-409d-bb39-e51257d377fd-->
<!--TEST RESULTS-->
<!--db3eaf1f-3910-43f1-83d2-587f415440d8-->
<!--TEST RESULTS-->
<!--d89bf233-45a5-4337-997f-7c0642eef5ca-->
<!--TEST RESULTS-->
<!--e470df70-a187-4341-806e-1b52a79b31f3-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67107">build#67107</a></summary>


| test_class                   | test_method                              | test_arguments                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:-----------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudArchiveRetentionTest    | test_delete                              | {"cloud_storage_type": 2, "retention_type": "retention.bytes"}                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba19-40d8-8e7e-0f4b69e51aab | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL           |
| CloudArchiveRetentionTest    | test_delete                              | {"cloud_storage_type": 2, "retention_type": "retention.ms"}                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba18-4afa-a719-83580f58de61 | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL            |
| CloudArchiveRetentionTest    | test_delete                              | {"cloud_storage_type": 1, "retention_type": "retention.bytes"}                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba18-4184-9a3d-5b7388128c29 | FLAKY         | 12/21    | upstream reliability is '100.0'. current run reliability is '57.14285714285714'. drift is 42.85714 and the allowed drift is set to 50. The test should PASS            |
| CloudArchiveRetentionTest    | test_delete                              | {"cloud_storage_type": 1, "retention_type": "retention.ms"}                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba19-46ea-ba8a-285fa383303d | FLAKY         | 9/21     | upstream reliability is '100.0'. current run reliability is '42.857142857142854'. drift is 57.14286 and the allowed drift is set to 50. The test should FAIL           |
| CloudRetentionTimelyGCTest   | test_retention_with_node_failures        | {"cloud_storage_type": 1}                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975adf-8206-4258-8f8c-99294c688974 | FLAKY         | 3/21     | upstream reliability is '100.0'. current run reliability is '14.285714285714285'. drift is 85.71429 and the allowed drift is set to 50. The test should FAIL           |
| CloudStorageTimingStressTest | test_cloud_storage                       | {"cleanup_policy": "compact,delete"}                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975adf-8205-450c-b0cb-b564be0cc4c2 | FLAKY         | 8/21     | upstream reliability is '100.0'. current run reliability is '38.095238095238095'. drift is 61.90476 and the allowed drift is set to 50. The test should FAIL           |
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves  | {"cleanup_policy": "compact,delete"}                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975adf-8206-4258-8f8c-99294c688974 | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS            |
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves  | {"cleanup_policy": "delete"}                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975adf-8207-41bf-8f48-e71e4cbf2487 | FLAKY         | 8/21     | upstream reliability is '100.0'. current run reliability is '38.095238095238095'. drift is 61.90476 and the allowed drift is set to 50. The test should FAIL           |
| ClusterHealthOverviewTest    | cluster_health_overview_baseline_test    |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975adf-8205-450c-b0cb-b564be0cc4c2 | FLAKY         | 20/21    | upstream reliability is '96.85990338164251'. current run reliability is '95.23809523809523'. drift is 1.62181 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest     | test_node_operations                     | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975adf-8207-41bf-8f48-e71e4cbf2487 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| ReadReplicasUpgradeTest      | test_upgrades                            | {"cloud_storage_type": 1}                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba19-40d8-8e7e-0f4b69e51aab | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| TestReadReplicaService       | test_identical_hwms                      | {"cloud_storage_type": 1, "partition_count": 5}                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975adf-8207-41bf-8f48-e71e4cbf2487 | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL           |
| TestReadReplicaService       | test_identical_hwms                      | {"cloud_storage_type": 1, "partition_count": 5}                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba18-4184-9a3d-5b7388128c29 | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL             |
| TestReadReplicaService       | test_identical_lwms_after_delete_records | {"cloud_storage_type": 1, "partition_count": 5}                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba19-46ea-ba8a-285fa383303d | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL             |
| TestReadReplicaService       | test_partition_movement                  | {"partition_count": 10}                                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba19-40d8-8e7e-0f4b69e51aab | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| TestReadReplicaService       | test_simple_end_to_end                   | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "partition_count": 10}                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975adf-8206-4258-8f8c-99294c688974 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 |
| TestReadReplicaService       | test_simple_end_to_end                   | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "partition_count": 10}                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba18-4afa-a719-83580f58de61 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 |
| TestReadReplicaService       | test_simple_end_to_end                   | {"cloud_storage_type_and_url_style": [1, "path"], "partition_count": 10}                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975adf-8207-41bf-8f48-e71e4cbf2487 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 |
| TestReadReplicaService       | test_simple_end_to_end                   | {"cloud_storage_type_and_url_style": [1, "path"], "partition_count": 10}                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba18-4184-9a3d-5b7388128c29 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 |
| TestReadReplicaService       | test_simple_end_to_end                   | {"cloud_storage_type_and_url_style": [1, "virtual_host"], "partition_count": 10}                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975adf-8205-450c-b0cb-b564be0cc4c2 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 |
| TestReadReplicaService       | test_simple_end_to_end                   | {"cloud_storage_type_and_url_style": [1, "virtual_host"], "partition_count": 10}                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba19-46ea-ba8a-285fa383303d | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 |
| TestReadReplicaService       | test_writes_forbidden                    | {"cloud_storage_type": 2, "partition_count": 10}                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba19-40d8-8e7e-0f4b69e51aab | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL            |
| TestReadReplicaService       | test_writes_forbidden                    | {"cloud_storage_type": 1, "partition_count": 10}                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba18-4afa-a719-83580f58de61 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL            |
| TimeQueryTest                | test_timequery                           | {"batch_cache": false, "cloud_storage": true, "spillover": true}                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba18-4184-9a3d-5b7388128c29 | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL             |
| TimeQueryTest                | test_timequery_with_spillover_gc_delayed |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67107#01975aef-ba19-46ea-ba8a-285fa383303d | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67165">build#67165</a></summary>


| test_class                | test_method                                         | test_arguments     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:----------------------------------------------------|:-------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest | test_read_when_segment_size_smaller_than_chunk_size |                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67165#01975e80-f871-4e35-b5a2-4b248c42c105 | FLAKY         | 20/21    | upstream reliability is '96.8'. current run reliability is '95.23809523809523'. drift is 1.5619 and the allowed drift is set to 50. The test should PASS                |
| MaintenanceTest           | test_maintenance_sticky                             | {"use_rpk": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67165#01975e7f-13ee-4836-8097-b2ef7ec0fef2 | FLAKY         | 18/21    | upstream reliability is '97.05882352941177'. current run reliability is '85.71428571428571'. drift is 11.34454 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67361">build#67361</a></summary>


| test_class                 | test_method                                            | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:---------------------------|:-------------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| PartitionReassignmentsTest | test_add_partitions_with_inprogress_reassignments      |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67361#01976fe1-6a73-4c47-af21-6aa9ec822f47 | FAIL          | 0/1      |          |
| TxUpgradeTest              | upgrade_does_not_change_tx_coordinator_assignment_test |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67361#01976fe1-6a73-4c47-af21-6aa9ec822f47 | FAIL          | 0/1      |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68146">build#68146</a></summary>


| test_class        | test_method         | test_arguments                                                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:------------------|:--------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| TieredStorageTest | test_tiered_storage | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "test_case": {"name": "(TS_Delete == True, SpilloverManifestUploaded == True, TS_Spillover_ManifestDeleted == True)"}} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68146#0197b691-e855-4fae-808f-17e234b52083 | FAIL          | 0/1      |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68635">build#68635</a></summary>


| test_class                              | test_method                | test_arguments                                                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------------------|:---------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterConfigTest                       | test_valid_settings        |                                                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68635#0197ef24-798e-48c5-a292-387cac1a056e | FLAKY         | 17/21    | upstream reliability is '83.65287588294652'. current run reliability is '80.95238095238095'. drift is 2.70049 and the allowed drift is set to 50. The test should PASS |
| InternalTopicProtectionLargeClusterTest | test_consumer_offset_topic |                                                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68635#0197ef23-ef8f-43b2-a4a9-afec5f68d410 | FLAKY         | 20/21    | upstream reliability is '99.71671388101983'. current run reliability is '95.23809523809523'. drift is 4.47862 and the allowed drift is set to 50. The test should PASS |
| TieredStorageTest                       | test_tiered_storage        | {"cloud_storage_type_and_url_style": [1, "path"], "test_case": {"name": "(TS_Delete == True, SpilloverManifestUploaded == True, TS_Spillover_ManifestDeleted == True)"}} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68635#0197ef23-ef8f-4299-8d70-5dca582f364a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68746">build#68746</a></summary>


| test_class        | test_method         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------|:--------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterConfigTest | test_valid_settings |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68746#0197f571-3b6e-43cc-b351-7c8d3c338af4 | FLAKY         | 18/21    | upstream reliability is '82.95668549905838'. current run reliability is '85.71428571428571'. drift is -2.7576 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68935">build#68935</a></summary>


| test_class           | test_method        | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------|:-------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RaftAvailabilityTest | test_one_node_down |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68935#01980a60-b919-4874-8f3d-5c7d20e7c077 | FLAKY         | 20/21    | upstream reliability is '96.91358024691358'. current run reliability is '95.23809523809523'. drift is 1.67549 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69385">build#69385</a></summary>


| test_class               | test_method           | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------|:----------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations  | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69385#01982f43-c412-4eee-888c-4ada9f59a271 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             |
| RandomNodeOperationsTest | test_node_operations  | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69385#01982f43-c415-4e02-8044-fc272faee553 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS            |
| ShardPlacementTest       | test_manual_rebalance |                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69385#01982f42-8461-44e0-a6c8-e5ffb534c038 | FLAKY         | 20/21    | upstream reliability is '99.7340425531915'. current run reliability is '95.23809523809523'. drift is 4.49595 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69426">build#69426</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69426#019831ff-d87d-474d-9b93-a916a898f90d | FLAKY         | 16/21    | upstream reliability is '96.61016949152543'. current run reliability is '76.19047619047619'. drift is 20.41969 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

Lazin committed:
archival: Use execution monitor in ntp_archiver

Use monitor to detect stalls and early exit from the background loop.
In case if the problem is detected print diagnostics and trace logs so
they would be visible on info level.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin commented:
upd: moved mutex into another PR

Lazin commented:
The PR https://github.com/redpanda-data/redpanda/pull/26412 was split from this one

dotnwat commented on src/v/ssx/BUILD:
```diff
@@ -120,6 +120,20 @@ redpanda_cc_library(
+redpanda_cc_library(
+    name = "execution_monitor",
+    hdrs = [
+        "execution_monitor.h",
+    ],
+    include_prefix = "ssx",
```
i guess we don't have a very specific definition of the scope of ssx, but this feels more like a utility than a seastar extension.

Lazin replied:
the scope is blurry, my hunch is that this is similar to watchdog and the watchdog is in ssx

dotnwat commented on src/v/ssx/execution_monitor.h:
```diff
@@ -0,0 +1,140 @@
+        _timer.set_callback([this,
+                             stall_cfg = std::move(stall_cfg),
+                             exit_cfg = std::move(exit_cfg)]() mutable {
+            auto h = _gate.hold();
```
what is this gate protecting? the units don't seem to be passed into any asychronous context.

Lazin replied:
this is correct, we don't need the gate
if the timer is disarmed it's guaranteed that the callback will not be called

dotnwat commented on src/v/ssx/execution_monitor.h:
```diff
@@ -0,0 +1,140 @@
+            if (stall_cfg.has_value()) {
+                bool has_stall = true;
+                basic_retry_chain_context<Clock>* last_ctx = nullptr;
+                for (retry_chain_context* ctx : stall_cfg.value().contexts) {
```
should last_ctx and ctx be the same type?

Lazin replied:
fixed

dotnwat commented on src/v/ssx/execution_monitor.h:
```diff
@@ -0,0 +1,140 @@
+                bool has_stall = true;
+                basic_retry_chain_context<Clock>* last_ctx = nullptr;
+                for (retry_chain_context* ctx : stall_cfg.value().contexts) {
+                    auto deadline = ctx->get_deadline(_stall_timeout);
```
it looks like this commit should come later in the series?

Lazin replied:
fixed

dotnwat commented on src/v/utils/retry_chain_node.h:
```diff
@@ -147,8 +148,83 @@
+#include <exception>
+#include <iterator>
 #include <variant>
 
+/// The context of the retry chain node.
+/// This object can collect traces in memory.
+/// It uses limited amount of memory which can be configured.
+/// It can also provide an abort source that can be used to
+/// stop the fiber.
+template<typename Clock = ss::lowres_clock>
+class basic_retry_chain_context {
+public:
+    /// Create a trace memory buffer with the specified size.
+    explicit basic_retry_chain_context(ss::sstring name, ss::abort_source& as, size_t size)
+      : _name(std::move(name))
+      , _as(as)
+      , _size_limit(size) {}
+
+    /// Add a trace to the buffer.
+    void add_trace(const ss::sstring& msg) const {
+        _last_trace_ts = Clock::now();
+        if (_buf.size() >= _size_limit) {
+            if (!_is_truncated) {
+                // If the buffer is full, we truncate it and set the flag.
+                _is_truncated = true;
+                fmt::format_to(std::back_inserter(_buf), "[truncated]");
+            }
+            return;
+        }
+        auto bii = std::back_inserter(_buf);
+        std::copy(msg.begin(), msg.end(), bii);
+        bii = '\n';
+    }
```
can you explain the semantics and motivation of those semantics of the logging in a class comment? for example, i thought we'd use a circular buffer so that when a stall occurred we'd see the last N messages. but this looks like once the buffer fills up we drop new messages. am i reading that correctly?

Lazin replied:
I updated the comment.
The idea is that this mechanism should be used for periodic but limited logging. You can't log infinite amount of log messages between resets. If the other type of trace collection is needed another type of  context should be added.

nvartolomei replied:
I think the value of this is very limited and the caller has to be very careful/conscious with buffer size/amount of logging. Maybe maintain a linked list of log lines with a total size? Would be pretty simple then to implement a fifo list of logs and it will be much easier to use/reason about.

Lazin replied:
I like linked lists but not as much

Lazin replied:
I replaced `fmt::memory_buffer` with the custom circular buffer implementation so now it is FIFO

dotnwat commented on src/v/ssx/execution_monitor.h:
```diff
@@ -0,0 +1,137 @@
+
+template<typename Clock = seastar::lowres_clock>
+using execution_monitor_stall_callback
+  = std::function<void(basic_retry_chain_context<Clock>&)>;
```
is the execution monitor specifically designed to work with the retry chain? if so, perhaps it should live in the same directory?

Lazin replied:
moved it to utils

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -147,8 +148,97 @@
+
+    /// Get the entire trace log as a string.
+    /// TODO: maybe add another method that allows to iterate over traces
+    ss::sstring get_trace_log() const { return {_buf.data(), _buf.size()}; }
```
This is a copy, right? Wouldn't a std::string_view be better?

Lazin replied:
this is used for testing, i'll add the comment

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -147,8 +148,97 @@
+    /// TODO: maybe add another method that allows to iterate over traces
+    ss::sstring get_trace_log() const { return {_buf.data(), _buf.size()}; }
+
+    void reset() {
```
Can you please document public methods? Especially the "suspended to" variables/methods.

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -147,8 +148,97 @@
+
+    ss::abort_source& as() noexcept { return _as; }
+
+    Clock::time_point get_reset_time() const noexcept { return _last_trace_ts; }
```
This is confusing naming. reset time sounds like the time when reset() was called but then we return `_last_trace_ts` which is also the time when last message was logged. Do you have a different intention for this method actually? Something like "last activity/checkpoint/heartbeat"?

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -147,8 +148,97 @@
+        _last_trace_ts = Clock::now();
+        if (_buf.size() >= _size_limit) {
+            if (!_is_truncated) {
+                // If the buffer is full, we truncate it and set the flag.
```
Comment is a bit misleading. Should be "we truncate" -> "we mark it as truncated"

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -147,8 +148,97 @@
+            return;
+        }
+        auto bii = std::back_inserter(_buf);
+        std::copy(msg.begin(), msg.end(), bii);
```
Should we check message length so it doesn't exceed the size? Pretty unlikely but accidents happen. No strong opinion there.

Lazin replied:
we will eventually exceed the size and stop appending data to the buffer

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -147,8 +148,97 @@
+
+    void suspend_to(Clock::time_point sp) noexcept { _suspended_to = sp; }
+
+    Clock::time_point get_deadline(Clock::duration stall_timeout) const {
```
This method name seems ambiguous to me. Especially with no code docs. It clearly is not a "deadline" in the usual sense I'm familiar with and instead it sounds like an "inactivity limit" of some sort.

I think this would benefit a better name and doc.

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -147,8 +148,97 @@
+class basic_retry_chain_context {
+public:
+    /// Create a trace memory buffer with the specified size.
+    explicit basic_retry_chain_context(ss::sstring name, ss::abort_source& as, size_t size)
```
what's the name for? code doc please

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -147,8 +148,97 @@
+template<typename Clock = ss::lowres_clock>
+class basic_retry_chain_context {
+public:
+    /// Create a trace memory buffer with the specified size.
```
> trace memory buffer

should we just say log memory buffer?

Lazin replied:
in the code below log messages are referred as 'traces'

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -374,7 +493,8 @@ class [[gnu::warn_unused]] basic_retry_chain_node {
     /// Deadline for retry attempts
     time_point _deadline;
     /// optional parent node or (if root) abort source for all fibers
```
outdated comment

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -435,6 +563,23 @@ class basic_retry_chain_logger final {
+    void bypass_tracing(Fn&& fn) {
+        std::exception_ptr err;
+        auto tmp = _has_tracing;
+        _has_tracing = false;
```
could be simpler with `ss::defer` to guarantee that has_tracing is reset? no strong opinion

Lazin replied:
IIRC defer allocates

nvartolomei replied:
nit: couldn't find any allocations; moreover it gets optimized down to way way less instructions

nvartolomei commented on src/v/utils/retry_chain_node.cc:
```diff
@@ -397,6 +456,16 @@ void basic_retry_chain_node<Clock>::check_abort() const {
+void basic_retry_chain_node<Clock>::maybe_add_trace(
+  const ss::sstring& s) const {
+    auto root = get_root();
+    if (!root->has_retry_chain_context()) {
```
You walk a "linked list" twice here. Maybe replace the `has_retry_chain_context` (which calls get_root too) together with std::get with a std::get_if?

Lazin replied:
it walks linked list once in `get_root`, then `has_retry_chain_context` is called on a root

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -435,6 +563,23 @@ class basic_retry_chain_logger final {
+    /// Invoke the lambda function but disable tracing while the function is
+    /// running.
+    template<class Fn>
+    void bypass_tracing(Fn&& fn) {
```
Wonder if this instead should be a function returning a "bypass_tracing_guard". Then it can be used in sync/async contexts too.

Can be done later.

Lazin replied:
It's not supposed to be used in async context, the only goal here is to avoid logging traces second time.

When the code detects stall and logs all accumulated traces it does this using the same logger. The logger normally would just add traces to the buffer once again. To avoid this this method was added.

nvartolomei commented on src/v/utils/execution_monitor.h:
```diff
@@ -0,0 +1,137 @@
+using execution_monitor_unexpected_shutdown_callback = std::function<void()>;
+
+/// The execution monitor is used to detect stalls in the execution
+/// of a process. It allows to register multiple callsites and checkpoint
```
Commit description and code comment is out of date/uses concepts which do not exist anymore/do not make sense.

callsite = context—I don't understand this
checkpoint = reset—Don't understand either

"when callsite is reached" what does this even mean if callsite is a context 🤷‍♂️

nvartolomei commented on src/v/utils/execution_monitor.h:
```diff
@@ -0,0 +1,137 @@
+/// The set of callsites is fixed at the construction time. The callsites
+/// are constructed as a set of objects and passed to the c-tor.
+template<typename Clock = seastar::lowres_clock>
+class basic_execution_monitor {
```
context_liveness_monitor?

nvartolomei commented on src/v/utils/execution_monitor.h:
```diff
@@ -0,0 +1,137 @@
+    void start(
+      std::optional<stall_detector_config> stall_cfg,
+      std::optional<unexpected_shutdown_detector_config> exit_cfg) {
+        if (!exit_cfg.has_value() && !stall_cfg.has_value()) {
```
why would one even call start with both of these empty? any reason these are not part of constructor(s)?

Lazin replied:
this is just a defensive check
they're not part of the c-tor because they're not used anywhere else

nvartolomei commented on src/v/utils/execution_monitor.h:
```diff
@@ -0,0 +1,137 @@
+    };
+
+    struct unexpected_shutdown_detector_config {
+        execution_monitor_unexpected_shutdown_callback cb;
```
some documentation on these members would help to understand the code much easier

i.e. cb // callback that will be called if the monitored gate does not have any holders
gate // gate to monitor

nvartolomei commented on src/v/utils/tests/execution_monitor_test.cc:
```diff
@@ -0,0 +1,225 @@
+            BOOST_FAIL("Stall detected at callsite: " + ctx.name());
+        },
+      .contexts = {}};
+    bool shutdown_detected = false;
```
missing assert on `shutdown_detected`?

Lazin replied:
fixed

Lazin committed:
utils: FIFO logging in retry_chain_context

Previously, the retry_chain_context used a buffer with limited size
which was able to accept new elements until it's full. After it become
full all new log trace messages were discarded.

This commit adds a circular_buffer implementation and uses it instead of
the old buffer. This makes the context evict messages in FIFO order. New
messages are always accepted but they may overwrite some old messages.
The first message can be overwritten partially. In order to avoid
confusion the '[truncated]' prefix is added to the log in this case.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -153,6 +153,163 @@
+    /// Return a view of all elements in the buffer in logical order (oldest to
+    /// newest)
+    auto view() {
+        return std::views::iota(size_t{0}, _size)
```
shouldn't this start at _start rather than 0?

Lazin replied:
no, it invokes `operator []` which starts from 0, the `_start` is accounted for by this operator

nvartolomei replied:
Ah. Nice.

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -153,6 +153,163 @@
+    auto view() {
+        return std::views::iota(size_t{0}, _size)
+               | std::views::transform(
+                 [this](size_t i) -> reference { return (*this)[i]; });
```
does this generate reasonable code?

Lazin replied:
what do you mean by "reasonable code"?

nvartolomei replied:
We're iterating on every char here. By reasonable code I mean whether using this method results in relatively lightweight iteration rather than calling function doing math and some other potential overhead for every char.

I wouldn't call this code reasonable in general case for this task https://godbolt.org/z/cjT7bqf4x but probably fine given that it is expected to be called relatively infrequently.

Well, same for push back. I assume that compiler can't optimize it to a simple memcpy.

Intuitively, linked list of messages would be more efficient all around. Or, if you opt for circular buffer with exact byte limits then with operations which are higher level (i.e. push_message) rather than push_back(char) where you can put things in the buffer with 2 memcpys and one adjustment to the size.

Lazin replied:
this was fixed

Lazin replied:
I didn't want to use linked list because of memory fragmentation etc. I also opted for a simple implementation so it will not overcomplicate this PR and will be easy to review.

nvartolomei commented on src/v/cluster/archival/ntp_archiver_service.cc:
```diff
@@ -382,6 +390,87 @@ ntp_archiver::ntp_archiver(
+void ntp_archiver::log_collected_traces() noexcept {
+    try {
+        _rtclog.bypass_tracing([this] {
+            _rtclog.info("Diagnostic dump start");
```
What about logging at error level so that this fails our CI tests? Or, shall we update ducktape regexp to catch these?

Lazin replied:
done

nvartolomei commented on src/v/utils/retry_chain_node.h:
```diff
@@ -153,6 +153,163 @@
+/// operations. When the buffer reaches its maximum capacity, new elements
+/// overwrite the oldest elements in FIFO order.
+template<typename T>
+class circular_buffer {
```
Needs a unit test.

nvartolomei commented on src/v/utils/tests/retry_chain_node_test.cc:
```diff
@@ -212,6 +212,17 @@ SEASTAR_THREAD_TEST_CASE(check_tracing) {
             != std::string_view::npos;
     BOOST_REQUIRE(found);
 
+    std::vector<ss::sstring> actual;
```
Looks like the meat of this commit was squashed into the first one. Squash then the test too otherwise the commit message is confusing and irrelevant.

nvartolomei commented on src/v/cluster/archival/ntp_archiver_service.cc:
```diff
@@ -631,7 +747,12 @@ ss::future<> ntp_archiver::sync_manifest_until_abort() {
-                co_await _leader_cond.wait();
+                constexpr auto leader_cond_timeout = 30s;
+                _rtctx.suspend_to(
+                  ss::lowres_clock::now() + leader_cond_timeout);
```
With very low probability but we have a potential race here where `now() + leader_cond_timeout` will match almost exactly with `_stall_timeout` periodic timer and we will have a chance of triggering the warning?

We should account for that to avoid noise.

oleiman commented on src/v/cluster/archival/ntp_archiver_service.cc:
```diff
@@ -1005,8 +1129,15 @@ ss::future<> ntp_archiver::upload_until_term_change_legacy() {
+            // Reset the context before sleeping to avoid holding to the memory.
+            // Most of the time the archiver spends in the following 'wait'
+            // call. It's not possible for it to get stalled there so it's safe
+            // to reset the context here. This guarantees that we only consuming
+            // additional memory for traces for the duration of the upload.
```
> consuming additional memory for traces

what stops us from accumulating trace logs while the context is suspended? or is this referring to the overhead of the actual trace dump?

oleiman replied:
actually, i guess this is a comment on releasing already produced traces, yeah?

Lazin replied:
It's needed to save memory. There is little value in holding to the trace log while sleeping here.

Lazin replied:
> actually, i guess this is a comment on releasing already produced traces, yeah?

yes

oleiman commented on src/v/utils/retry_chain_node.h:
```diff
@@ -154,6 +154,211 @@
+        _total_added++;
+    }
+
+    void push_back(std::string_view sv) {
```
feels a bit odd having this interface bound to `basic_string_view<char>` when the container is generic. could remove the top-level template and call out that it's for signed chars only, or restrict the template to char types and have this function take `basic_string_view<T>`, or have this function take a generic range.

Lazin replied:
fixed

oleiman commented on src/v/utils/retry_chain_node.h:
```diff
@@ -154,6 +154,199 @@
+
+private:
+    std::vector<char> _data;
+    size_t _capacity;
```
nitpick: could be const

Lazin replied:
default assignment operator will be impossible with const

oleiman commented on src/v/utils/retry_chain_node.h:
```diff
@@ -154,6 +154,199 @@
+
+    /// Add an element to the buffer. When the buffer is full, the oldest
+    /// element is overwritten.
+    void push_back(char value) {
```
nitpick:

```suggestion
    void push_back(value_type value) {
```

Lazin replied:
fixed

oleiman commented on src/v/utils/retry_chain_node.h:
```diff
@@ -154,6 +154,199 @@
+    }
+
+private:
+    std::vector<char> _data;
```
nitpick: 

```suggestion
    std::vector<value_type> _data;
```

Lazin replied:
done

oleiman commented on src/v/utils/retry_chain_node.h:
```diff
@@ -154,6 +154,199 @@
+                std::memcpy(
+                  _data.data(), sv.data() + space_available, remaining);
+                _start = remaining % _capacity;
+            }
```
do we really need the branches and bare `memcopy` here? e.g.

```suggestion
            size_t n = std::min(space_available, bytes_to_copy);
            std::copy(sv.begin(), std::next(sv.begin(), n), std::back_inserter(_data));
            std::copy(std::next(sv.begin(), n), sv.end(), _data.begin());
            _size += n;
            _start = (bytes_to_copy - n) % _capacity;
```

probably wind up with roughly comparable codegen in release mode. idk, take it or leave it :)

Lazin replied:
yep, added that

oleiman commented on src/v/utils/tests/retry_chain_node_test.cc:
```diff
@@ -245,18 +243,182 @@ SEASTAR_THREAD_TEST_CASE(check_tracing) {
+    }
+    BOOST_REQUIRE_EQUAL(result, "redpanda");
+
+    buf.push_back(std::string_view{"head"});
+    // this is supposed to hit case two: not enough space, need two memcpy calls
+    buf.push_back(std::string_view{"1234567"});
```
should there be another check for `buf` contents here? overwritten and so on?

Lazin replied:
right

oleiman reviewed (commented):
few minor suggestions on the circular buffer, preceding commits lgtm
